Python/fileutils.c

```diff
@@ -2096,6 +2107,124 @@ _Py_abspath(const wchar_t *path, wchar_t **abspath_p)
 #endif
 }
 
+// The Windows Games API family does not provide these functions
+// so provide our own implementations. Remove them in case they get added
+// to the Games API family
+// Note that this implementation does not handle all the same cases as the real
+// function, but we expect games are very unlikely to encounter the more obscure
+// cases.
+#if defined(MS_WINDOWS_GAMES) && !defined(MS_WINDOWS_DESKTOP)
+HRESULT
+PathCchSkipRoot(const wchar_t *path, const wchar_t **rootEnd)
+{
+    if (path[0] == '\\') {
+        /* relative path with root e.g. \Windows */
+        if (path[1] != '\\') {
+            *rootEnd = path + 1;
+            return S_OK;
+        }
+
+        /* UNC drives e.g. \\server\share or \\?\UNC\server\share */
+        const wchar_t *end = path + 2;
+        if (!wcsnicmp(end, L"?\\UNC\\", 6)) {
+            end += 6;
+        }
+
+        end = wcschr(end, '\\');
+        if (!end) {
+            *rootEnd = path + wcslen(path);
+            return S_OK;
+        }
+        end = wcschr(end + 1, '\\');
+        *rootEnd = (!end) ? path + wcslen(path) : end + 1;
+        return S_OK;
+    }
+    /* absolute / relative path with drive, e.g. C: or C:\ */
+    else if (isalpha(path[0]) && path[1] == ':') {
+        *rootEnd = (path[2] == '\\') ? path + 3 : path + 2;
+        return S_OK;
+    }
+
+    /* relative path */
+    return E_INVALIDARG;
+}
+
+static HRESULT
+PathCchStripToRoot(wchar_t *path, size_t size)
+{
+    wchar_t *end;
+    if (PathCchSkipRoot(path, &end) == S_OK) {
+        if (*end == '\0') {
+            return S_FALSE;
+        }
+        *end = '\0';
+    }
+
+    return E_INVALIDARG;
+}
+
+static wchar_t*
+PathAddBackslashW(wchar_t *path)
+{
+    size_t len;
+    if (!path) {
+        return NULL;
+    }
+    len = wcslen(path);
+    if (len && path[len - 1] != '\\') {
+        path[len++] = '\\';
+        path[len] = '\0';
+    }
+    return path + len;
+}
+
+#ifndef PATHCCH_ALLOW_LONG_PATHS
+#define PATHCCH_ALLOW_LONG_PATHS 0x01
+#endif
+
+static HRESULT
+PathCchCombineEx(wchar_t *buffer, size_t bufsize, const wchar_t *dirname,
```

Comment:
The call shouldn't fail with insufficient buffer error when `relfile` is a rooted path, and `bufsize` is big enough for the root of `dirname` plus `relfile`, but not for all of `dirname`. In this case, the effective value of `dir_len` can be calculated via `PathCchSkipRoot(dirname, &end)` and, if the latter succeeds, `dir_len = (size_t)(end - dirname)`. Otherwise use `dir_len = wcslen(dirname)`. Copy it via `wcsncpy(buffer, dirname, dir_len)`. Then there's no need to implement `PathCchStripToRoot()`.